kong/clustering/update_config.lua

```diff
@@ -0,0 +1,240 @@
+      local keys = new_tab(count, 0)
+      local i = 0
+      for k in pairs(value) do
+        i = i + 1
+        keys[i] = k
+      end
+
+      sort(keys)
+
+      local o = new_tab(count, 0)
+      for i = 1, count do
+        o[i] = keys[i] .. ":" .. to_sorted_string(value[keys[i]])
+      end
+
+      value = concat(o, ";", 1, count)
+
+      return #value > 512 and ngx_md5_bin(value) or value
+    end -- isarray
+
+  end   -- table
+
+  error("invalid type to be sorted (JSON types are supported)")
+end
+
+local function calculate_config_hash(config_table)
+  if type(config_table) ~= "table" then
+    local config_hash = ngx_md5(to_sorted_string(config_table))
+    return config_hash, { config = config_hash, }
+  end
+
+  local routes    = config_table.routes
+  local services  = config_table.services
+  local plugins   = config_table.plugins
+  local upstreams = config_table.upstreams
+  local targets   = config_table.targets
+
+  local routes_hash    = routes    and ngx_md5(to_sorted_string(routes))    or DECLARATIVE_EMPTY_CONFIG_HASH
+  local services_hash  = services  and ngx_md5(to_sorted_string(services))  or DECLARATIVE_EMPTY_CONFIG_HASH
+  local plugins_hash   = plugins   and ngx_md5(to_sorted_string(plugins))   or DECLARATIVE_EMPTY_CONFIG_HASH
+  local upstreams_hash = upstreams and ngx_md5(to_sorted_string(upstreams)) or DECLARATIVE_EMPTY_CONFIG_HASH
+  local targets_hash   = targets   and ngx_md5(to_sorted_string(targets))   or DECLARATIVE_EMPTY_CONFIG_HASH
+
+  config_table.routes    = nil
+  config_table.services  = nil
+  config_table.plugins   = nil
+  config_table.upstreams = nil
+  config_table.targets   = nil
+
+  local config_hash = ngx_md5(to_sorted_string(config_table) .. routes_hash
+                                                             .. services_hash
+                                                             .. plugins_hash
+                                                             .. upstreams_hash
+                                                             .. targets_hash)
+
+  config_table.routes    = routes
+  config_table.services  = services
+  config_table.plugins   = plugins
+  config_table.upstreams = upstreams
+  config_table.targets   = targets
+
+  return config_hash, {
+    config    = config_hash,
+    routes    = routes_hash,
+    services  = services_hash,
+    plugins   = plugins_hash,
+    upstreams = upstreams_hash,
+    targets   = targets_hash,
+  }
+end
+
+local hash_fields = {
+    "config",
+    "routes",
+    "services",
+    "plugins",
+    "upstreams",
+    "targets",
+  }
+
+local function fill_empty_hashes(hashes)
+  for _, field_name in ipairs(hash_fields) do
+    hashes[field_name] = hashes[field_name] or DECLARATIVE_EMPTY_CONFIG_HASH
+  end
+end
+
+function _M.new(conf)
+  local self = {
+    declarative_config = declarative.new_config(conf),
+  }
+
+  return setmetatable(self, _MT)
+end
+
+function _M:execute(config_table, config_hash, hashes)
```

Comment:
I suggest not holding states with an "update" object, and the name "execute" is too generic. This should be a bare function. You could make it a function named `update_config`, taking `cert_cfg` and `declarative_config` as parameters rather than bundle those fields without clear relation.
Or if you prefer, you could have a "self" parameter, and then bind this function to data plane objects.

Reply:
rename to `config_helper`, it contains two config related functions.